### PR TITLE
Nicer Errors

### PR DESCRIPTION
**Pull request by stefanpenner**
@jfelchner note: I'll be updating this initial comment with new details around the error page as they arise.  Full work required to get to this state will be updated in https://github.com/ember-cli/ember-cli/issues/6441

# Default View

Only necessary information is expanded

ember-cli-error-page-light-collapsed

# Fully Expanded Information

(with both banners visible as well)

ember-cli-error-page-light

I'd love it if anyone and everyone could give us some feedback on this.

# Goals

My main goal with this is to give the developer actionable information in a quickly parsable way.

I especially have an eye on newbies who will have no idea what's going on when they get here.

## Friendly and Inviting

Hitting this page is not an ideal outcome so we want to make sure that the page itself is friendly and not something that you loathe seeing.

## Professional and Confident

We want users who come to this page to know that even though something went wrong, EmberCLI has their back.

## Easily Parseable Title

The current build errors are auto-generated by Broccoli and don't contain the phrasing the makes it as obvious as it could be exactly what went wrong.

## Easily Locatable Location

The filename, line and location of the error need to be very easy to find.
The line and location should be "click to copy" to make it easy to paste into your editor or elsewhere.

## Easy to Grok Code

The actual code snippet should be syntax highlighted, include auto-generated line numbers, and highlight the line of the error, so that it's easy to see at a glance what needs to be corrected.

## Hide It Till You Need It

There are lots of thing we could show, but (especially for newbies) we need to only show information that users are going to want 80% of the time.

## Broccoli Pipeline Graph

One of the most difficult things to grok is what order Broccoli is parsing your tree in.  And when you have a problem, you're often not sure if it was actually the node that it errored on or a previous node that did something weird.

We'll display a partial graph which will show the current (errored) node as well as the inputs to that node.

## Stacktraces that Make Sense

Stacktraces need to be color-coded and have their information aligned so that they are easy to scan.

## Mobile Responsive

When you're working in Simulator or are developing on a phone or tablet, the error page should still provide actionable information in an easy to use way.

### Upgrade Banner

Allow the user to understand that a new EmberCLI version is available which may resolve their issue.

### Auto-Post Banner

We'll use some simple heuristics to determine whether the build error is the result of an EmberCLI bug or something the user did.  If it's the former, we can auto-post an issue with detailed information in exactly the way the EmberCLI maintainers would like it.  All the user has to do is hit "Create Issue".

### Links to the Community

For newbies, they may not know all the resources available to them to get access to the great Ember community, so we'll call out that if they need more help, StackOverflow, Slack and Github are available.

### Additional Information Icons

For the newbies we'll add some "Additional Information" icons to each of the sections which can link somewhere to help them understand what the information in each of the sections represents.

### CODE REVIEW

File: lib/middleware.js [hunk 1]
```diff
@@ -115,10 +114,24 @@ module.exports = function getMiddleware(watcher, options) {
...
+          file: buildError.file,
+          extension: path.extname(buildError.file).replace('.', ''),
+          treeDir: buildError.treeDir,
+        }
+      };
+
       response.setHeader('Content-Type', 'text/html')
       response.writeHead(500)
+
+var errorTemplate = handlebars.compile(fs.readFileSync(path.resolve(__dirname, 'templates/error.html')).toString())
```
Comment (stefanpenner) [inline lib/middleware.js]: hack, so I could iterate on the style without having to reload the server all the time.

**Comment by jfelchner** [comment]
/cc @stefanpenner

Removed this comment in favor of updating the initial comment in this thread for easier readability.

**Comment by jfelchner** [comment]
Just so everyone understands those are flat mockups, not an implemented page.  Some of those things may not be possible.

**Comment by Turbo87** [comment]
whoa, that looks really awesome!! :+1:

**Comment by jfelchner** [comment]
At a later date I'd like for that "Partial" link on the graph to expand the graph to show all the nodes in the build.

**Comment by kellyselden** [comment]
@jfelchner Amazing work! One thought is how many things are in your way before you get to what you want, which is the error and/or stacktrace. Maybe the first two banners of new version and bug post should be at the bottom? They would help newbies, but get in the way of power users. Just an observation. Thoughts.

**Comment by rwjblue** [comment]
Pretty exciting!

**Comment by josemarluedke** [comment]
Woow, this would be awesome! I really like the first design by @jfelchner. Not sure how I few about the banner to post a new issue. That would probably cause several people to just click and open issues about their application error without actually be related to Ember CLI. But besides that, this looks amazing.

**Comment by jfelchner** [comment]
Thanks for he great feedback everyone!

Let me explain in further detail about the "auto-posting" banner since a lot of people seem interested in that. 

This isn't going to be a feature just for newbies, I see everyone using this. The idea behind it is that it will rarely show up, but we'll do some simple heuristics based on the error that will show the banner _only if_ it's a type of error that the EmberCLI core team thinks is likely a bug with CLI itself (and not something like a Sass parsing issue for example)

So as far as the maintainers and the burden on them are concerned, they would want users to post these issues. Obviously this is a thing that will require some tweaking to get just right, but the hope is that in the end it provides a means to give CLI maintainers much more actionable information than what they currently receive in order to make CLI even more stable.

**Comment by jfelchner** [comment]
@kellyselden I really like the banners at the top and as I mentioned in the previous comment, the auto-post is something that should rarely show. However I agree with your observation about the version banner. Especially for apps hat are on an older version of CLI with no plans to upgrade.

Perhaps @stefanpenner knows if we can add a build flag in `ember-cli-build` that could be used to modify the error page and simply always disable that banner?

**Comment by stefanpenner** [comment]
- [ ] Error: Parse error on line 2 likely needs to be towards the top
- [ ] Error location is currently a region, but doesn't clearly share where in that region this issue occurred `

**Comment by locks** [comment]
(I'm fixing the rebase mess I made)

**Pull request closed without merging**